[홈 화면 (전 화면에 확대 적용할 예정)] 파이어베이스 AuthState 및 네트워크 연결 상태 변화 처리 구현

jungkeunking opened the pull request:
- [x] 화면별 UiState들의 상위 타입인 BaseUiState 정의해 uid 및 isOnline 프로퍼티 오버라이딩해 사용
- [x] 프로필 화면을 위해 확장 함수로 정의해둔 getUidCallbackFlow를 AuthStateRepository의 메서드로 활용
- [x] 뷰모델에서 uid 및 isOnline 상태 변화를 감지해 UiState를 업데이트해주는 확장 함수들 정의
- [x] Firebase AuthState가 변화한 시점에 uid 값이 null인 경우, 로그인 화면으로 이동시키는 확장 함수 정의
- [x] 네트워크 연결 상태가 해제된 시점에 필요한 연산을 함수 타입으로 갖는 확장 함수 정의
- [x] 위의 두 확장 함수를 홈 화면에 적용해, uid 값이 null일 때 로그인 화면으로 이동하고, 오프라인일 때 심벌 및 플랜 마커(닫힌 보물상자) 클릭 비활성화 구현
- [x] 프로필 이미지 프리로드 함수를 추상 클래스 메서드에서 확장 함수로 변경 (해당 브랜치 작업 내용과 무관하나, 작업하면서 상속에 대해 더 찾아보다가 알게 된 내용을 적용했으니 양해 부탁드립니다!)

** 추후: 홈 화면 이외의 다른 화면에도 각각 상태 변화 처리 적용할 예정

** AuthStateChange를 감지해야 하는 이유에 대해 공식문서에 나와 있는 내용 참고 링크:
https://firebase.google.com/docs/auth/android/manage-users?_gl=1*rg8h5s*_up*MQ..*_ga*MTg2OTc0MDc2NS4xNzExNjE2MDA1*_ga_CW55HF8NVT*MTcxMTYxNjAwNS4xLjAuMTcxMTYxNjAwNS4wLjAuMA..
근데 해당 내용이 최신화됐는지 의문이라 좀 더 찾아봐야 할 거 같습니다
다른 데서 본 바로는 파이어베이스 최신 버전부터 token state change는 감지하지 못해서 idTokenListener인가 사용해야 된다고 봤습니다

** 아 그리고 스낵바 메시지 가운데 정렬을 못 하겠는데 아시는 분 도움 주시면 감사하겠습니다!

** 추가 수정 사항:
- 로그인 표기가 Login, LogIn 뒤섞여 있는 걸 Login으로 통일
- BaseUiState에 에러메시지 id 프로퍼티 추가 및 홈 화면에서 보물상자 마커 불러오는 함수의 오류 처리
에러 핸들링을 넘어가지 말라고 권하는 코틀린 언어 개발자의 글 참고:
https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07
뷰모델에서 에러를 uiState에 넘기는 방식의 예시 참고:
https://developer.android.com/topic/architecture/ui-layer
- 홈 화면에서 네트워크 연결 해제 후 재연결시, 비활성화 기능을 재활성화
- null uid 값일 때 로그인 화면으로 이동하는 글로벌 액션을 백스택 클리어하도록 수정
- 홈 화면 및 뷰모델 전체적인 리팩토링:
- 1) 인스턴스가 관리할 필요 없는 상태들을 로컬 변수로 전환
- 2) 홈 프래그먼트가 비대해져 네이버 맵과 관련된 함수들을 OnMapReadyCallback 인터페이스의 확장 함수로 전환
- 3) 기록 상세 화면에서 기록 삭제후 홈 화면에서 삭제된 마커 처리 구현을 내비게이션을 활용해 변경 (삭제된 기록의 id를 전달받아 직접 마커를 숨김 처리 -> popUpTo를 통해 뷰모델의 UiState를 리셋)
- 4) 홈 뷰모델에서 로컬 DB에서 장소 데이터를 불러와 마커로 변환하는 작업의 로직 수정 (Visit, Plan 마커들을 따로 처리 및 map 함수를 활용해 플로우에 바로 연산)

jungkeunking committed:
refactor: preloadProfileImage 함수를 AbstractFragment 메서드 대신 확장 함수로 변경

jungkeunking committed:
feat: Firebase Auth 변화를 관찰해 uid 값을 갱신하는 레포지토리 추가 및 네트워크 연결 레포지토리 일부 수정

jungkeunking committed:
feat: 홈 화면에서 uid가 null로 변할 때 로그인 화면으로 이동 구현

jungkeunking committed:
remove: 불필요한 클래스, 생성자 파라미터, 임포트 제거

jungkeunking committed:
refactor: 홈 화면에서 uid null 처리 로직들 제거

setSymbolClick 함수 내부의 Firebase.auth.signOut()는 테스트용으로 작성한 코드를 실수로 이전 커밋에 포함시켜서 삭제

jungkeunking committed:
feat: 홈 화면에서 온라인 연결 해제시 스낵바 메시지 표시 및 심벌, 플랜 클릭 비활성화를 일괄 처리

jungkeunking commented on app/src/main/java/com/treasurehunt/ui/home/HomeViewModel.kt:
```diff
@@ -67,32 +68,33 @@ class HomeViewModel @Inject constructor(
-
-        _uiState.update {
-            it.copy(uid = uid)
+    private fun updateUid() {
```
init 블록에 인라인으로 바로 호출하는 게 낫겠네요

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/model/BaseUiState.kt:
```diff
@@ -0,0 +1,6 @@
+package com.treasurehunt.ui.model
+
+abstract class BaseUiState(
+    open val uid: String?,
+    open val isOnline: Boolean
+)
```
abstract class로 만들어 모든 UiState에서 사용하게 하는 방법 좋은 것 같습니다!

juhan2103 reviewed (commented):
@jungkeunking 정근님 구현 수고 많으셨습니다! 온라인 연결 해제 시 스낵바 메세지 뜨는거 확인했고 정렬에 대해서는 저도 방법을 찾을 수가 없네요 ㅜㅜ

snackbar material design link: https://github.com/material-components/material-components-android/blob/master/docs/components/Snackbar.md

이건 materail design snackbar 링크입니다!

jungkeunking commented:
> https://github.com/material-components/material-components-android/blob/master/docs/components/Snackbar.md

오 안 그래도 제 아이폰 앱들 확인해보니까, 사파리, 네이버 지도 등등 저런 스낵바 뜨길래, 비슷하게 하면 좋겠다고 생각했습니다!
걔네들도 매터리얼 사용하나 보네요

jungkeunking committed:
fix + refactor: LogIn -> Login 이름 수정 및 홈 화면 리팩토링

jungkeunking committed:
feat: BaseUiState에 에러메시지 id 프로퍼티 추가 및 홈 화면에서 보물상자 마커 불러오는 함수의 오류 처리

에러 핸들링을 넘어가지 말라고 권하는 코틀린 언어 개발자의 글 참고:
https://elizarov.medium.com/kotlin-and-exceptions-8062f589d07
뷰모델에서 에러를 uiState에 넘기는 방식의 예시 참고:
https://developer.android.com/topic/architecture/ui-layer

jungkeunking committed:
fix: 홈 화면에서 네트워크 연결 해제 후 재연결시, 비활성화 기능을 재활성화시키지 않은 오류 수정

jungkeunking committed:
fix + refactor: 지도 다이얼로그에서 UiState의 uid 기본값이 null이어서 홈으로 이동해버리는 오류 수정 및 리팩토링

jungkeunking committed:
feat: 홈 화면에서 오프라인 상태에서 마커 클릭시 오류 메시지 표시 및 기록 삭제 후 홈 화면으로 돌아와 마커 삭제 처리 로직 수정 (popUpTo 활용)

jungkeunking committed:
fix + refactor: 로그인 화면으로 이동하는 글로벌 액션을 백스택 클리어하도록 수정 및 홈 화면과 뷰모델 리팩토링

donghyun81 commented on app/src/main/java/com/treasurehunt/util/Extensions.kt:
```diff
@@ -69,3 +137,13 @@ fun Fragment.preloadProfileImage(storageUrl: String?) {
+            }
+        }
+    }    
+}
```
uiState에서 uid를 추적해서 값이 없을 경우에 로그인 화면으로 이동 할 수 있는 기능을 여러 프래그먼트에서 사용 할 수 있도록 확장 함수로 구현한 부분 기발하고 좋았습니다.

donghyun81 commented on app/src/main/java/com/treasurehunt/util/AuthStateRepositoryImpl.kt:
```diff
@@ -0,0 +1,30 @@
+            Firebase.auth.removeAuthStateListener(callback)
+        }
+    }
+}
```
firebase api를 통해서 일일히 호출하는것 보다 repository를 통해 프로젝트에서 관리해서 데이터의 역할에 맞게 가공할 수 있고 ui 패키지에서 분리 되는 부분이 좋은것 같습니다.

jungkeunking replied:
의도를 정확하게 보셨네요~~
감사합니다

donghyun81 reviewed (commented):
구현 고생 많으셨습니다. 네트워크 대응을 확장 할 수 있도록 구현한 부분 기발하였던것 같습니다!

jungkeunking committed:
refactor: 홈 화면 리팩토링

글라이드 콜백 리스너를 반환하는 인스턴스 함수를 동반 객체의 프로퍼티로 변경, 코루틴 및 플로우 코드 간결화 등

jungkeunking committed:
refactor: 홈 화면에서 리플렉션 활용, 코루틴 최적화, 로직 중복 제거

jungkeunking committed:
fix: 홈 화면에서 네트워크 변화 처리시, UiState의 false 기본값 때문에 onLost 블록이 실행되는 오류 수정

jungkeunking committed:
refactor: 홈 화면에서 함수 타입 활용 코드 간결화 및 맵다이얼로그 화면에서 비즈니스 로직 분리

onConnectivityChange의 아규먼트 간결화, 맵다이얼로그의 비즈니스 로직을 프래그먼트 및 홈 뷰모델에서 분리, 원격 DB에 plan 삽입시 로컬 max id를 활용해 업데이트 과정 생략, 비즈니스 로직 처리를 뷰에서 뷰모델 코루틴으로 이전 등

jungkeunking committed:
feat + refactor: BaseUiState에 errorMessageResId -> loadStatus 프로퍼티로 변경 및 홈 화면에서 로딩 상태 처리

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/home/MapDialogViewModel.kt:
```diff
@@ -0,0 +1,84 @@
+     * MUST BE updated if it changes in the future
+     * */
+    fun insertPlan(place: PlaceModel): Job = viewModelScope.launch {
+        val localPlaceMaxId = placeRepo.getLocalPlaceMaxId().first()
+        val remotePlanId = placeRepo.insert(
+            place.asPlaceDTO(localId = localPlaceMaxId + 1)
+        )
```
혹시 이 부분은 왜 getLocalPlaceMaxId라는 함수를 만들어서 사용하신걸까요? 업데이트 생략때문에 사용한다고 써있는 것 같긴 한데 어떻게 업데이트를 생략할 수 있는지를 모르겠습니다!

jungkeunking replied:
이건 원래 함수랑 비교해보면 알 수 있습니다
이전의 MapDialogFragment의 아래 함수가 전신인데요
    private suspend fun insertPlan(place: PlaceModel): String {
        val remotePlanId = viewModel.insertPlace(
            place.asPlaceDTO()
        )
        val localPlanId = viewModel.insertPlace(
            place.asPlaceEntity(remoteId = remotePlanId)
        )
        viewModel.updatePlace(
            remotePlanId,
            place.asPlaceDTO(localId = localPlanId)
        )
        return remotePlanId
    }
보시다시피 위의 연산 과정을 정리하면,
원격DB에 추가 -> 로컬 DB에 추가 -> 원격DB에 로컬 아이디로 업데이트
이렇게 됩니다
왜냐면 최초에 원격 푸시할 때는 로컬 아이디가 없으니까요
지금 연산 과정은 이렇게 됩니다
로컬DB의 아이디 계산 -> 원격DB에 추가 -> 로컬 DB에 추가
원격 DB를 조작하는 연산 수를 하나 줄이고 로컬 DB 쿼리 연산을 하나 늘린 셈인데, 전자는 원격 통신이 요구되므로 더 효율적이라고 판단했습니다
그리고 레코드를 수정하는 연산보다 테이블의 값 하나를 읽어오는 연산이 훨씬 간단하지 않을까 하는 개인적인 판단입니다

juhan2103 reviewed (commented):
@jungkeunking 정근님 구현 수고하셨습니다! 코드리뷰 하나 달아놨으니 답변 해주시면 감사하겠습니다!
